base/src/main/java/proguard/classfile/util/inject/CodeInjector.java

```diff
@@ -0,0 +1,216 @@
+import proguard.classfile.attribute.Attribute;
+import proguard.classfile.attribute.CodeAttribute;
+import proguard.classfile.attribute.visitor.AllAttributeVisitor;
+import proguard.classfile.attribute.visitor.AttributeNameFilter;
+import proguard.classfile.attribute.visitor.AttributeVisitor;
+import proguard.classfile.editor.CodeAttributeEditor;
+import proguard.classfile.editor.InstructionSequenceBuilder;
+import proguard.classfile.instruction.Instruction;
+import proguard.classfile.util.ClassUtil;
+import proguard.classfile.util.InternalTypeEnumeration;
+import proguard.classfile.util.inject.argument.InjectedArgument;
+import proguard.classfile.util.inject.location.InjectStrategy;
+
+/**
+ * This utility class allow for injecting a method invocation instruction optionally with arguments
+ * modeled by instances of classes implementing {@link InjectedArgument} interface to the specified
+ * target method at an offset determined by the implementation of the {@link InjectStrategy}
+ * interface.
+ *
+ * <p>Example usage: new CodeInjector() .injectInvokeStatic(logUtilClass, logDebugMethod, new
+ * ConstantPrimitive<Integer>(1), new ConstantString("Hello world")) .into(MainProgramClass,
+ * mainMethod) .at(new FirstBlock()) .commit();
+ *
+ * @author Kymeng Tang
+ */
+public class CodeInjector {
+  private List<ClassMethodPair> targets;
+  private ClassMethodPair content;
+  private InjectStrategy injectStrategy;
+  private List<InjectedArgument> arguments = new ArrayList<>();
+
+  /**
+   * Specify the static method to be invoked.
+   *
+   * @param clazz The class in which the static method belongs to.
+   * @param method The method to be invoked.
+   */
+  public CodeInjector injectInvokeStatic(Clazz clazz, Method method) {
+    assert content == null
+        : "The injection content: `"
+            + renderInjectionContent(content.clazz, content.method, arguments)
+            + "` "
+            + "has already been specified.";
+
+    assert (method.getAccessFlags() & AccessConstants.STATIC) != 0
+            && !method.getName(clazz).equals(ClassConstants.METHOD_NAME_CLINIT)
+        : "The method to be injected must be a (non-class initializer) static method.";
+
+    content = new ClassMethodPair(clazz, method);
+    return this;
+  }
+
+  /**
+   * Specify the static method to be invoked.
+   *
+   * @param clazz The class in which the static method belongs to.
+   * @param method The method to be invoked.
+   * @param arguments a list of arguments to be passed to the method to be invoked.
+   */
+  public CodeInjector injectInvokeStatic(
+      Clazz clazz, Method method, InjectedArgument... arguments) {
+    injectInvokeStatic(clazz, method);
+
+    InternalTypeEnumeration parametersIterator =
+        new InternalTypeEnumeration(method.getDescriptor(clazz));
+    Iterator<InjectedArgument> argumentsIterator = Arrays.stream(arguments).iterator();
+
+    while (parametersIterator.hasNext() || argumentsIterator.hasNext()) {
```

Comment:
```suggestion
    while (parametersIterator.hasNext() && argumentsIterator.hasNext()) {
```
Shouldnt this be an && to be safe? If somehow we erroneously have a difference in the size of parameters and arguments the next call would fail. Maybe we can properly handle that though.